gnome-image-installer/pages/install/gis-scribe.c

```diff
@@ -1150,6 +1186,128 @@ gis_scribe_begin_verify (GisScribe *self,
   return g_object_ref (gpg_stdin);
 }
 
+static void
+checksum_in_thread (GTask        *task,
+                    gpointer      source,
+                    gpointer      task_data,
+                    GCancellable *cancellable)
+{
+  GisScribe *self = source;
+  GisScribeChecksumData *checksum_data = task_data;
+  gsize len;
+  guint8 buf[BUFFER_SIZE];
+  g_autoptr(GChecksum) sha256sum = g_checksum_new (G_CHECKSUM_SHA256);
+  g_autoptr(GError) error = NULL;
+  const gchar *digest;
+
+  for (;;) {
+    if (!g_input_stream_read_all (checksum_data->input, buf, sizeof (buf), &len,
+                                  NULL, &error))
+      {
+        task_return_error (self, task, g_steal_pointer (&error));
+        return;
+      }
+
+    if (len == 0)
+      break;
+
+    g_checksum_update (sha256sum, buf, len);
+  }
+
+  digest = g_checksum_get_string (sha256sum);
+  if (g_strcmp0 (digest, checksum_data->expected_checksum) != 0)
+    {
+      task_return_new_error (
+          self, task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("Image checksum ‘%s‘ does not match expected checksum ‘%s‘."),
+          digest, checksum_data->expected_checksum);
+      return;
+    }
+
+  g_task_return_boolean (task, TRUE);
+}
+
+static GOutputStream *
+gis_scribe_begin_verify_checksum (GisScribe           *self,
+                                  GCancellable        *cancellable,
+                                  GAsyncReadyCallback  callback,
+                                  gpointer             data)
+{
+  g_autoptr(GTask) task = g_task_new (self, cancellable, callback, data);
+  GisScribeChecksumData *task_data = g_slice_new0 (GisScribeChecksumData);
+  g_autofree gchar *checksum_path = g_file_get_path (self->checksum);
+  g_autofree gchar *checksum_contents = NULL;
+  gsize checksum_len;
+  gchar *cur;
+  gint pipefd[2];
+  g_autoptr(GError) error = NULL;
+
+  g_task_set_source_tag (task, GUINT_TO_POINTER (GIS_SCRIBE_TASK_VERIFY));
+  g_task_set_task_data (task, task_data,
+                        (GDestroyNotify) gis_scribe_checksum_data_free);
+
+  if (!g_file_query_exists (self->checksum, NULL))
+    {
+      task_return_new_error (
+          self, task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("The checksum file ‘%s’ does not exist."),
+          checksum_path);
+      return NULL;
+    }
+
+  /* Read in the checksum file to get the expected checksum. */
+  if (!g_file_load_contents (self->checksum, cancellable, &checksum_contents,
+                             &checksum_len, NULL, &error))
+    {
+      task_return_error (self, task, g_steal_pointer (&error));
```

Comment:
I think this is an existing bug, but currently this will deadlock the installer using `task_return_error` because the `gis_scribe_begin_*` functions used for task setup are called with the mutex locked. I briefly changed them back to `g_task_return_error`, but then the error message that shows up is wrong because it doesn't set `GisScribe::error`. I'm not sure what the best approach is.

Reply:
Ah, I'd forgotten this particular wart. I was never too happy with how the subtasks' errors propagate upwards. I'm not sure if there is more advanced technology for multiplexed GTask these days.

Perhaps the different fields could be guarded by different mutexes? Looking at the code, I think `gis_scribe_write_thread_await_gpg()` could be made to only access `self->outstanding_tasks`…

Reply:
You could also change the `outstanding_tasks` usage to atomic integer access, but I'm not sure that's the right model since the top level seems like it wants to keep the lock until all the tasks are setup.

Reply:
Oh, that could work. I think I wanted to have the property that it would always be non-zero at the end of this function, so that one of the earlier subtasks failing couldn't cause the task we're setting up to be completed while we're still going.

I don't actually know if that's possible. And it could be faked by adding an extra field to the bitfield for "SETUP".

Reply:
I ended up just locking and unlocking the mutex around each access to `outstanding_tasks` in `gis_scribe_write_async`. Initially I thought it was important that the mutex was held until all the subtasks were started, but I don't think that's necessary. I added tests for this (and manually tested by running the installer with no signature present) and it did the right thing.

Reply:
OK, makes sense!